src/core/server/config/config_service.mock.ts

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { BehaviorSubject } from 'rxjs';
+import { ObjectToConfigAdapter } from './object_to_config_adapter';
+
+import { ConfigService } from './config_service';
+
+type MethodKeysOf<T> = {
+  [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never
+}[keyof T];
+
+type PublicMethodsOf<T> = Pick<T, MethodKeysOf<T>>;
+
+type ConfigSericeContract = PublicMethodsOf<ConfigService>;
+const createConfigServiceMock = () => {
+  const mocked: jest.Mocked<ConfigSericeContract> = {
+    atPath: jest.fn(),
+    getConfig$: jest.fn(),
+    optionalAtPath: jest.fn(),
+    getUsedPaths: jest.fn(),
+    getUnusedPaths: jest.fn(),
+    isEnabledAtPath: jest.fn(),
+  };
+  mocked.atPath.mockReturnValue(new BehaviorSubject({}));
+  mocked.getConfig$.mockReturnValue(new BehaviorSubject(new ObjectToConfigAdapter({})));
+  mocked.getUsedPaths.mockResolvedValue([]);
+  mocked.getUnusedPaths.mockResolvedValue([]);
+  mocked.isEnabledAtPath.mockResolvedValue(true);
+  return mocked;
+};
+
+export const configServiceMock = {
+  create: createConfigServiceMock,
```

Comment:
Rather than exporting a top level `configServiceMock` property that consuming code invokes a function from, why not just export the functions themselves? I suspect each mock will only have a small number of exports, maybe 3 on average.

I don't mind boilerplate, so I don't have strong opinions, but since we're establishing a new convention here, I figured it'd be good to document the rationale.

Reply:
I found it a bit harder to read when we have additional helpers
```js
import { createLoggingServiceMock, collectLoggingServiceMock } from '../logging_service.mock'
const logger = createLoggingServiceMock();
expect(collectLoggingServiceMock(logger)).toMatchSnapshot(...)
// vs
import { loggingServiceMock } from '../logging_service.mock'
const logger = loggingServiceMock.create();
expect(loggingServiceMock.collect(logger)).toMatchSnapshot(...)
```
bit this is a matter of opinion. I'm fine with either option

Reply:
Works for me